Added skiplist implementation of member, redefine skipper and add ele…

vladi-tech opened the pull request:
…ment

# Description
Added implementation of Skip List.
Added functions:
            1)member, which returns a pointer to the element with max value <= of the key
            2)redefine skipper, which changes the skip values after adding a new element
            3)add element, which adds the element in order (sorted from lowest to highest value)

# Changes
None

# Checklist
Make sure all checks are passed

- [ x] Code is in the proper folder
- [ x] Proper label is added
- [ x] At least one approving review is present

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.cpp:
```diff
@@ -0,0 +1,299 @@
+    {
+        curr->skip = nullptr;
+
+        if (counter % skipValue == 0 || !curr->next)
```
Ако края не е на това отстояние, ще го хванеш ли?

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.cpp:
```diff
@@ -0,0 +1,299 @@
+template <class T>
+SkipList<T>::SkipList()
+{
+    start = nullptr;
+    end = nullptr;
```
Инициализиращ списък

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.cpp:
```diff
@@ -0,0 +1,299 @@
+}
+
+template <class T>
+size_t SkipList<T>::size() const
```
Става ли да не обхождаме списъка всеки път, когато трябва да вземе размера?

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.cpp:
```diff
@@ -0,0 +1,299 @@
+template <class T>
+Node<T> *SkipList<T>::member(const T &key) const ///Returns the member with the max value <= of the one we are searching for, aka the member after the new Value should be added
+{
+    if (start->value > key) ///If start has a higher value, we return nullptr, as the element with value key will be added before start
```
Мога ли да нямам start?

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.cpp:
```diff
@@ -0,0 +1,299 @@
+
+    Node<T> *curr = start;
+
+    while (curr->skip->value <= key && curr) ///We skip most of the elements, as the list is already sorted when we add a new one
```
На доста места забелязвам, че изпускаш моментите, в които указателя, на който разчиташ може да го няма

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.cpp:
```diff
@@ -0,0 +1,299 @@
+}
+
+template <class T>
+void SkipList<T>::printSkips() const
```
Why????

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.h:
```diff
@@ -0,0 +1,42 @@
+#include <iostream>
+
+template <class T>
+struct Node
```
Дай да я вкараме вътре в класа

NickGeorgiev commented on group4/82019_Vladimir_Toshev/skiplist.cpp:
```diff
@@ -0,0 +1,299 @@
+    }
+}
+
+int main()
```
Това няма място тук

NickGeorgiev reviewed (changes requested):
1. Иска ми се да му напишеш някой друг unit тест
2. member може да го преименуваме на locate и да напишеш един member който връща истина или лъжа
3. Като е шаблон или вкарай всичко в хедъра или включи кода накрая на хедъра и го защити от повторно включване

vladi-tech commented:
Description
Added the requested changes to the skiplist template, as well as the unit tests and makefile.

Changes
1) Moved the files to  a new folder named "SkipList".
2) Renamed the locate function.
3) Added function bool member.
4)Added unit tests.

Checklist
Make sure all checks are passed

[ x] Code is in the proper folder
[ x] Proper label is added
[ x] At least one approving review is present

NickGeorgiev commented on group4/82019_Vladimir_Toshev/SkipList/skiplist.h:
```diff
@@ -0,0 +1,216 @@
+
+    bool member(const T &key) const
+    {
+        if (!this->size)
```
Ако ще е число, нека да не разчитаме, че ще се конвертира като bool

NickGeorgiev commented on group4/82019_Vladimir_Toshev/SkipList/skiplist_tests.cpp:
```diff
@@ -0,0 +1,398 @@
+#include "skiplist.h"
+#include "../../../shared/specs/specHelper.h"
```
Идеята на makefile-а е, че си спестяваш ужасния релативен път. `shared/specs/specHelper.h` е достатчно, за да компилираш през makefile-a

NickGeorgiev reviewed (changes requested):
Когато адресираш коментари, недей да качваш файловете наново, понеже така реално правя втори път ревю на едно и също нещо и не виждам експлицитно какво точно си променил. Променяй текущите фейлове, за да се виждат промените в commit-ите

vladi-tech commented:
Description
Added the requested changes to the skiplist member function, as well as removed the path to shared specs in the  unit tests file.

Changes
1)We now use operator == to check if the size is 0.
2)The long path to shared specs was removes.

Checklist
Make sure all checks are passed

[ x] Code is in the proper folder
[ x] Proper label is added
[ x] At least one approving review is present
